libs/common/src/platform/state/implementations/default-active-user-state.ts

```diff
@@ -1,71 +1,142 @@
     private chosenStorageLocation: AbstractStorageService & ObservableStorageService,
   ) {
-    this.formattedKey$ = this.accountService.activeAccount$.pipe(
-      map((account) =>
-        account != null && account.id != null
-          ? userKeyBuilder(account.id, this.keyDefinition)
-          : null,
-      ),
-      tap(() => {
-        // We have a new key, so we should forget about previous update promises
-        this.updatePromise = null;
+    this.activeUserId$ = this.accountService.activeAccount$.pipe(
+      // We only care about the UserId but we do want to know about no user as well.
+      map((a) => a?.id),
+      // To avoid going to storage when we don't need to, only get updates when there is a true change.
+      distinctUntilChanged(),
+    );
+
+    const userChangeAndInitial$ = this.activeUserId$.pipe(
+      // If the user has changed, we no longer need to lock an update call
+      // since that call will be for a user that is no longer active.
+      tap(() => (this.updatePromise = null)),
+      switchMap(async (userId) => {
+        // We've switched or started off with no active user. So,
+        // emit a fake value so that we can fill our share buffer.
+        if (userId == null) {
+          return FAKE;
+        }
+
+        const fullKey = userKeyBuilder(userId, this.keyDefinition);
+        const data = await getStoredValue(
+          fullKey,
+          this.chosenStorageLocation,
+          this.keyDefinition.deserializer,
+        );
+        return [userId, data] as CombinedState<T>;
+      }),
+    );
+
+    const latestStorage$ = this.chosenStorageLocation.updates$.pipe(
+      // Use withLatestFrom so that we do NOT emit when activeUserId changes because that
+      // is taken care of above, but we do want to have the latest user id
+      // when we get a storage update so we can filter the full key
+      withLatestFrom(this.activeUserId$),
+      switchMap(async ([storageUpdate, userId]) => {
+        if (userId == null) {
+          // Technically we could return undefined here and everything should work
+          // because the above observable should already be producing FAKE
+          // when there is no active user so this will likely just overwrite that.
+          // So as long as we return a value that doesn't make it's way to the consumer
+          // (undefined or FAKE), we are good.
+          return FAKE;
+        }
+
+        const fullKey = userKeyBuilder(userId, this.keyDefinition);
+        if (storageUpdate.key !== fullKey) {
+          // This is an update that is not applicable to us,
+          // return undefined and filter out undefined below
+          // don't return FAKE here because we don't want this
+          // scenario to fill up the replay subjects buffer
+          return undefined;
+        }
+
+        // We can shortcut on updateType of "remove"
+        // and just emit null. Currently, "remove" is rarely,
+        // if ever, called.
```

Comment:
```suggestion
        // and just emit null. 
```

I don't think the frequency is relevant